nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullHandler.java

```diff
@@ -0,0 +1,218 @@
+
+      state.reportMatch(
+          analysis
+              .getErrorBuilder()
+              .createErrorDescription(
+                  new ErrorMessage(ErrorMessage.MessageTypes.POSTCONDITION_NOT_SATISFIED, message),
+                  tree,
+                  buildDescriptionFromChecker(tree, analysis),
+                  state));
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * All overriding methods can only strengthen the postcondition of its super method. All
+   * overriding methods can only add new field names to the set of fields of its super method
+   * mentioned in {@code EnsuresNonNull}.
+   */
+  @Override
+  protected void validateOverridingRules(
+      Set<String> overridingFieldNames,
+      NullAway analysis,
+      VisitorState state,
+      MethodTree tree,
+      Symbol.MethodSymbol overriddenMethod) {
+    Set<String> overriddenFieldNames =
+        ContractUtils.getFieldNamesFromAnnotation(overriddenMethod, annotName);
+    if (overriddenFieldNames == null) {
+      return;
+    }
+    if (overridingFieldNames == null) {
+      overridingFieldNames = Collections.emptySet();
+    }
+    if (overridingFieldNames.containsAll(overriddenFieldNames)) {
+      return;
+    }
+    overriddenFieldNames.removeAll(overridingFieldNames);
+
+    StringBuilder errorMessage = new StringBuilder();
+    errorMessage.append(
+        "postcondition inheritance is violated, this method must guarantee that all fields written in overridden method @EnsuresNonNull annotation are @NonNull at exit point as well. Fields [");
+    Iterator<String> iterator = overriddenFieldNames.iterator();
+    while (iterator.hasNext()) {
+      errorMessage.append(iterator.next());
+      if (iterator.hasNext()) {
+        errorMessage.append(", ");
+      }
+    }
+    errorMessage.append(
+        "] must explicitly appear as parameters at this method @EnsuresNonNull annotation");
+    state.reportMatch(
+        analysis
+            .getErrorBuilder()
+            .createErrorDescription(
+                new ErrorMessage(
+                    ErrorMessage.MessageTypes.WRONG_OVERRIDE_POSTCONDITION,
+                    errorMessage.toString()),
+                tree,
+                buildDescriptionFromChecker(tree, analysis),
+                state));
+  }
+
+  /**
+   * On every method annotated with {@link com.uber.nullaway.qual.EnsuresNonNull}, this method,
+   * injects the {@code Nonnull} value for the class fields given in the {@code @EnsuresNonNull}
+   * parameter to the dataflow analysis.
+   */
+  @Override
+  public NullnessHint onDataflowVisitMethodInvocation(
+      MethodInvocationNode node,
+      Types types,
+      Context context,
+      AccessPathNullnessPropagation.SubNodeValues inputs,
+      AccessPathNullnessPropagation.Updates thenUpdates,
+      AccessPathNullnessPropagation.Updates elseUpdates,
+      AccessPathNullnessPropagation.Updates bothUpdates) {
+    if (node.getTree() == null) {
```

Comment:
Actually I don't know how that might happen. The method returns ```@Nullable```, I wanted to make sure that I won't have NPEs in the code. Is it possible that it returns ```null``` when processing bytecodes ?

Reply:
I think it can only return `null` if it's some kind of synthetic node that the Checker Framework inserted during CFG construction.  It doesn't hurt to have a null check here, maybe we can just add a comment as to why we need it.